lib/ClangToAst.ml

```diff
@@ -1396,6 +1482,63 @@ let rec translate_stmt (env : env) (s : Clang.Ast.stmt_desc) : Krml.Ast.expr =
     end
   | If { cond = { desc = BinaryOperator { lhs; kind = NE; rhs }; _ }; then_branch; _ }
     when has_pointer_type lhs && is_null rhs -> translate_stmt env then_branch.desc
+
+  | If { cond; then_branch; else_branch; _ } when is_tag_check env cond ->
+      let var, variant, case_ref = deconstruct_tag_check env cond in
+
+      let lid = Helpers.assert_tlid var.typ in
+      let case, fs = match LidMap.find_opt lid !type_def_map with
+        | Some (lazy (Variant branches)) -> List.nth branches variant
+        | _ -> fatal_error "Expected a tagged union expression"
+      in
+
+      let name, case_t = match fs with
+        | [(n, (t, _))] -> n, t
+        | _ -> failwith "Tagged union case has more than one field"
+      in
+
+      let binder = Helpers.fresh_binder name case_t in
+      let pat = Krml.Ast.with_type case_t (PBound 0) in
+      (* We need to add the new binder to the environment before translating
+         the branches.
+         We also need to be able to later retrieve the content of the variant
+         constructor when accessing the corresponding field.
+         To do so, we store in the environment a variable called v!!{atom},
+         where {atom} is replaced by the binder's atom.
+         We then update the environment for `var` to store that it is a tagged
+         union in case `case`, and that the corresponding content is ``v!!{atom}`.
+         Importantly, `v!!{atom}` is not a valid variable name, and therefore does
+         not conflict with existing variables.
+      *)
+      let env_binder_name = binder.node.name ^ "!!" ^ show_atom_t binder.node.atom in
+      let new_env = add_var env (env_binder_name, case_t) in
+
+      (* We only change the state of the tagged union case to translate the if branch,
+         which is the one where we checked the tag of the variable *)
+      (* TODO: Should we sanity-check that old is None? As, if we are already in
+         a tagged union case, there is no need for rechecking the tag? *)
+      let old = !case_ref in
+      case_ref := Some { case; var = env_binder_name };
+      let then_e = translate_stmt new_env then_branch.desc in
+      case_ref := old;
+
+      let else_e = match else_branch with
+        | None -> Helpers.eunit
+        (* We translate the else branch with the old environment, without
+           adding a binder for the pattern *)
+        | Some el -> translate_stmt env el.desc
+      in
+      let t =
+        match then_e.typ, else_e.typ with
+        | TAny, t -> t
+        | t, TAny -> t
+        | _ -> then_e.typ
+      in
+
+      let then_branch = [binder], Krml.Ast.with_type t (PCons (case, [pat])), then_e in
+      let else_branch = [], Krml.Ast.with_type t PWild, else_e in
+      Krml.Ast.with_type t (EMatch (Unchecked, var, [then_branch; else_branch]))
```

Comment:
does there need to be an optimization pass later on on Ast that collapses nested matches into a flat one?

Reply:
It's a possibility for sure, however I'd be in favor of waiting until we have more examples that exhibit this pattern. I am not yet sure whether examples will typically pattern-match on the tag (in this case, they might rather use a switch?), or just extract it in one specific union case, with a tag sanity-check as an if/then/else